### ISSUE TITLE
Cannot call pure function on `before_expiry(result)`

### ISSUE THREAD

**Issue by JonasAlaif** [viperproject/prusti-dev#900]
The following code:
```rust
use prusti_contracts::*;
fn main() {}
#[pure]
fn foo(_x: &i32) -> bool { true }

#[after_expiry(
    foo(before_expiry(result))
)]
fn id(x: &mut i32) -> &mut i32 { x }
```
Results in the error:
```
The required permission Acc(old[l2](_0.val_ref).val_ref.val_int, read) cannot be obtained.
```
I suspect that it is due to an incorrect encoding of the argument `before_expiry(result)`, where the encoding should instead be `old[l2](_0.val_ref).val_int` (without the extra `.val_ref`).

c.f. the `package[l2]` below:
![prusti_error](https://user-images.githubusercontent.com/13405042/157730840-20b9d66f-264f-4e0a-be84-b47f20f06023.svg)

**Comment by JonasAlaif** [comment]
In this case the issue can be solved by changing `foo(before_expiry(result))` to `before_expiry(foo(result))`. However, this doesn't work when we want to call the function with other arguments, e.g. the very common case of:
```rust
use prusti_contracts::*;
fn main() {}

struct Container(i32);
impl Container {
  #[pure]
  fn foo(&self, x: &i32) -> bool { self.0 == *x }
  
  #[after_expiry(
      self.foo(before_expiry(result))
  )]
  fn get(&mut self) -> &mut i32 { &mut self.0 }
}
```

**Comment by Pointerbender** [comment]
I've got another case for this one using slices:

```rust
use prusti_contracts::*;

#[trusted]
#[requires(to_slice.len() == from_slice.len())]
#[ensures(*to_slice == *old(from_slice))]
fn copy_slice(to_slice: &mut [u8], from_slice: &[u8]) {
    to_slice.copy_from_slice(from_slice);
}

pub struct A {
    a: [u8; 16],
}

impl A {
    #[pure]
    pub const fn get(&self) -> &[u8] {
        &self.a
    }

    #[ensures(result.len() == 16)]
    #[after_expiry(self.get() == before_expiry(&*result))]
    pub fn get_mut(&mut self) -> &mut [u8] {
        &mut self.a
    }
}

#[requires(from_slice.len() == 16)]
#[ensures(a.get() == old(from_slice))]
pub fn test(a: &mut A, from_slice: &[u8]) {
    let to_slice = a.get_mut();
    copy_slice(to_slice, from_slice);
}
```

But with a slightly different error message:

```
[2022-05-04T14:08:23Z INFO  prusti_viper::encoder::encoder] Encoding: prusti_example::A::get_mut (prusti_example::{impl#0}::get_mut)
error: [Prusti internal error] Prusti encountered an unexpected internal error
  --> src/lib.rs:91:5
   |
91 | /     pub fn get_mut(&mut self) -> &mut [u8] {
92 | |         &mut self.a
93 | |     }
   | |_____^
   |
   = note: We would appreciate a bug report: https://github.com/viperproject/prusti-dev/issues/new
   = note: Details: cannot generate fold-unfold Viper statements. The required permission Pred(old[l4](_0.val_ref), read) cannot be obtained.
           

[2022-05-04T14:08:23Z INFO  prusti_viper::encoder::encoder] Encoding: prusti_example::test (prusti_example::test)
error: [Prusti internal error] Prusti encountered an unexpected internal error
   --> src/lib.rs:98:1
    |
98  | / pub fn test(a: &mut A, from_slice: &[u8]) {
99  | |     let to_slice = a.get_mut();
100 | |     copy_slice(to_slice, from_slice);
101 | | }
    | |_^
    |
    = note: We would appreciate a bug report: https://github.com/viperproject/prusti-dev/issues/new
    = note: Details: cannot generate fold-unfold Viper statements. The required permission Pred(_old$l2$0, read) cannot be obtained.
```

**Comment by JonasAlaif** [comment]
Fixed by #1014
